src/BuildPrediction/Predictors/MasmItemsPredictor.cs

```diff
@@ -0,0 +1,131 @@
+        /// optional name or location of the browse information file.
+        /// </summary>
+        internal const string BrowseFileMetadata = "BrowseFile";
+
+        /// <summary>
+        /// Item meta data name that determines if the item group should skipped, and not assembled.
+        /// </summary>
+        internal const string ExcludedFromBuildMetadata = "ExcludedFromBuild";
+
+        /// <summary>
+        /// Item meta data name that represents the fully qualified path to the assembler output.
+        /// </summary>
+        internal const string ObjectFileNameMetadata = "ObjectFileName";
+
+        /// <summary>
+        /// Item meta data name that represents the paths for the assembler to search for include file(s).
+        /// </summary>
+        internal const string IncludePathsMetadata = "IncludePaths";
+
+        /// <summary>
+        /// Character separator for values specified in %(MASM.IncludePaths).
+        /// </summary>
+        private static readonly char[] IncludePathsSeparator = { ';' };
+
+        /// <inheritdoc />
+        public void PredictInputsAndOutputs(ProjectInstance project, ProjectPredictionReporter reporter)
+        {
+            // This is based on $(VCTargetsPath)\BuildCustomizations\masm.targets, if the before targets
+            // property isn't set then the masm.targets haven't been imported, and no @(MASM) items will
+            // be processed by the MASM targets.
+            if (string.IsNullOrWhiteSpace(project.GetPropertyValue(MasmBeforeTargetsPropertyName)))
+            {
+                return;
+            }
+
+            ICollection<ProjectItemInstance> masmItems = project.GetItems(MasmItemName);
+            if (masmItems.Count == 0)
+            {
+                return;
+            }
+
+            var reportedIncludes = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            foreach (ProjectItemInstance masmItem in masmItems)
+            {
+                if (IsExcludedFromBuild(masmItem))
+                {
+                    continue;
+                }
+
+                ReportInputs(reporter, masmItem, reportedIncludes);
+                ReportOutputs(reporter, masmItem);
+            }
+        }
+
+        private bool IsExcludedFromBuild(ProjectItemInstance masmItem) =>
+            masmItem.GetMetadataValue(ExcludedFromBuildMetadata).Equals(bool.TrueString, StringComparison.OrdinalIgnoreCase);
+
+        private void ReportInputs(ProjectPredictionReporter reporter, ProjectItemInstance masmItem, HashSet<string> reportedIncludes)
+        {
+            reporter.ReportInputFile(masmItem.EvaluatedInclude);
+
+            string[] includePaths = masmItem.GetMetadataValue(IncludePathsMetadata)
+                .Split(IncludePathsSeparator, StringSplitOptions.RemoveEmptyEntries);
+
+            // Avoid reporting paths that we've already reported for this project.
+            foreach (string includePath in includePaths.Except(reportedIncludes))

```

Comment:
Line can be expensive. Instead try something like...

```cs
foreach (string includePath in includePaths)
{
    // Avoid reporting paths that we've already reported for this project.
    if (reportedIncludes.Add(includePath))
    {
        reporter.ReportInputDirectory(includePath);
    }
}
```

Reply:
Sure